### PR TITLE
structured logging

### PR DESCRIPTION
**Pull request by pohly**
This lays the foundation and converts the operator as a smaller piece of code to structured logging.

Both pmem-csi-driver and pmem-csi-operator now have a `-logging-format=json|text` command line parameter.

TODO:
- [x] ~~support choosing the logging format of the operator when deploying through OLM~~ - cannot be implemented at the moment, operator defaults to plain text
- [x] support choosing the logging format of the driver when deploying through the operator
- [x] admin documentation

### CODE REVIEW

**Comment by mythi** [comment]
@pohly  have you checked what improvements `controller-runtime` v0.7* brings for this? the manager carries a logger and it can be used from controllers via context.

**Comment by pohly** [comment]
I've not checked what logger controller-runtime offers. Ultimately we want consistent logging PMEM-CSI, whether it is in the driver (no controller-runtime) and in the operator (with controller-runtime), so we have to ensure that whatever logging calls are coming from controller-runtime also go to klog.

FWIW, I've pushed the commit which enables `-logging-format=json` with zapper as formatter - it uses the same code as Kubernetes.

**Comment by pohly** [comment]
Unfortunately https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/log/log.go uses its own methods for getting a logger from the context, which makes it incompatible with https://github.com/go-logr/logr/blob/4fa77cb7175ca349567983fd9cc9fd94462f41fa/logr.go#L197-L222

Looks like the default in controller-runtime is to not log anything: https://github.com/kubernetes-sigs/controller-runtime/blob/f3831c24e5e99c242f993fde8aad18047ca47570/pkg/log/log.go#L46-L49

That can and should be changed.

**Comment by mythi** [comment]
> uses its own methods for getting a logger from the context, 

AFAICS the commit timestamps indicate controller-runtime changes were made first

> Looks like the default in controller-runtime is to not log anything:

kubebuilder scaffolding sets `ctrl.SetLogger(zap.New(zap.UseDevMode(true)))` but klog can also be used.

**Comment by pohly** [comment]
> kubebuilder scaffolding sets ctrl.SetLogger(zap.New(zap.UseDevMode(true))) but klog can also be used.

We have neither of that in PMEM-CSI. Now I wonder how useful the logging would have been to debug issues...

**Comment by pohly** [comment]
> AFAICS the commit timestamps indicate controller-runtime changes were made first

Yes. I think logr reacted to other people introducing their own custom solutions.

I filed an issue against controller-runtime and it seems that this can be resolved quickly: https://github.com/kubernetes-sigs/controller-runtime/issues/1275

File: pkg/pmem-csi-operator/main.go [hunk 1]
```diff
@@ -37,17 +39,39 @@ func printVersion() {
 	klog.Info(fmt.Sprintf("Go OS/Arch: %s/%s", runtime.GOOS, runtime.GOARCH))
 }
 
-var driverImage string
+var (
+	driverImage = flag.String("image", "", "docker container image used for deploying the operator.")
+	logFormat   = func() *flagLogOptions {
+		f := &flagLogOptions{}
+		flag.Var(f, "logging-format", "determines log output format")
```
Comment (avalluri) [inline pkg/pmem-csi-operator/main.go:46]: What are the possible format options, and how to use them.

File: pkg/pmem-csi-operator/controller/deployment/controller_driver.go [hunk 2]
```diff
@@ -185,70 +192,38 @@ func (op *ObjectPatch) Apply(c client.Client, labels map[string]string) error {
...
-		}
-		allObjects = append(allObjects, o)
-		if o, err = d.redeployControllerDriver(r); err != nil {
-			return fmt.Errorf("failed to update controller driver: %v", err)
-		}
-		allObjects = append(allObjects, o)
-		if o, err = d.redeployNodeDriver(r); err != nil {
-			return fmt.Errorf("failed to update node driver: %v", err)
+
+		for name, handler := range subObjectHandlers {
```
Comment (avalluri) [inline pkg/pmem-csi-operator/controller/deployment/controller_driver.go:221]: These changes are not related logging, would be nice to keep them a separate commit/PR.

Comment (pohly) [reply pkg/pmem-csi-operator/controller/deployment/controller_driver.go:221]: While it would be nicer, that implies doing the code cleanup work twice because the change was done in parallel to adding ctx (basically I noticed that I would have to make the same change multiple times and instead decided to reduce the amount of similar functions) and now does not apply to "devel".

I don't think the "nicer" part is worth the extra effort.

**Review by avalluri — APPROVED** [review]

**Comment by avalluri** [comment]
Test failure:
```
[2020-12-15T10:38:44.022Z] /tmp/go-build557579245/b528/pmem-csi-driver.test flag redefined: log_dir
[2020-12-15T10:38:44.022Z] panic: /tmp/go-build557579245/b528/pmem-csi-driver.test flag redefined: log_dir
[2020-12-15T10:38:44.022Z] 
[2020-12-15T10:38:44.022Z] goroutine 1 [running]:
[2020-12-15T10:38:44.022Z] flag.(*FlagSet).Var(0xc00012a120, 0x19b9800, 0x23ada90, 0x17e11f0, 0x7, 0x1817b09, 0x2f)
[2020-12-15T10:38:44.022Z] 	/go/src/flag/flag.go:851 +0x485
[2020-12-15T10:38:44.022Z] flag.(*FlagSet).StringVar(...)
[2020-12-15T10:38:44.022Z] 	/go/src/flag/flag.go:754
[2020-12-15T10:38:44.022Z] k8s.io/klog/v2.InitFlags(0x0)
[2020-12-15T10:38:44.022Z] 	/mnt/workspace/gopath/pkg/mod/k8s.io/klog/v2@v2.3.0/klog.go:425 +0x7f
[2020-12-15T10:38:44.022Z] github.com/intel/pmem-csi/pkg/pmem-csi-driver.init.2()
```

**Comment by pohly** [comment]
I pushed a version which has `--logging-format` parameters in pmem-csi-operator and pmem-csi-driver and an API option for it in v1beta1. That part is ready for review and merging.

The actual *output* however via the forked klogr is sub-optimal: because it always has to serialize itself due to the lack of a `klog.InfoSDepth`, we get a single "msg" value even when the caller has used structured logging.

We need https://github.com/kubernetes/klog/pull/197 before we can do proper structured, contextual logging in PMEM-CSI. Or fork klog...

**Comment by pohly** [comment]
This is based on top of PR #830. Let's merge that one first.

**Comment by pohly** [comment]
> The actual output however via the forked klogr is sub-optimal: because it always has to serialize itself due to the lack of a klog.InfoSDepth, we get a single "msg" value even when the caller has used structured logging.

I think we should merge this PR now and improve the output later. It's a step in the right direction and avoids the overhead of rebasing each time we make conflicting changes.

@avalluri: agreed? Please review.

File: pkg/apis/pmemcsi/v1beta1/deployment_types.go [hunk 3]
```diff
@@ -50,6 +50,15 @@ const (
 	DeviceModeFake DeviceMode = "fake"
 )
 
+type LogFormat string
+
+const (
+	// LogFormatText selects logging via the traditional glog (aka klog) plain text format.
+	LogFormatText LogFormat = "text"
+	// DeviceModeText selects logging via the klog JSON format.
```
Comment (avalluri) [inline pkg/apis/pmemcsi/v1beta1/deployment_types.go:58]: nit: DeviceModeText -> LogFormatText

File: test/e2e/operator/deployment_api.go [hunk 4]
```diff
@@ -247,6 +248,24 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
...
+			deployment.Spec.Image = ""
+			deployment.Spec.PMEMPercentage = 50
+			deployment.Spec.LogFormat = api.LogFormatJSON
+
+			deployment = deploy.CreateDeploymentCR(f, deployment)
+			defer deploy.DeleteDeploymentCR(f, deployment.Name)
+
+			validateDriver(deployment)
+
+			output, err := e2epod.GetPodLogs(f.ClientSet, "default", deployment.Name+"-controller-0", "pmem-driver")
```
Comment (avalluri) [inline test/e2e/operator/deployment_api.go:262]: Can you please use `d.Namespace` instead of "default", that way this test will work with #829 without and further chagnes.

Comment (pohly) [reply test/e2e/operator/deployment_api.go]: Done.

It also occurred to me that I should be testing *all* log format options (default, text, json), so I added that, too.

File: pkg/apis/pmemcsi/v1beta1/deployment_types.go (hunk 3 above)
Comment (pohly) [reply pkg/apis/pmemcsi/v1beta1/deployment_types.go]: Fixed.

**Review by avalluri — APPROVED** [review]
